2022-04-06.VPN-Kit.md

```diff
@@ -0,0 +1,18 @@
+---
+updated: 2022-04-06
+authors:
+- name: Dave Scott
+  uri: https://github.com/djs55
+subject: 'A MirageOS VPN Kit'
+permalink: 2022-04-06.VPN-Kit
+---
+
+# A MirageOS VPN Kit
+
+[Recently, I posted](https://www.docker.com/blog/how-docker-desktop-networking-works-under-the-hood/) about how [VPN Kit](https://github.com/moby/vpnkit), built with MirageOS libraries, powers [Docker Desktop](https://www.docker.com/products/docker-desktop). Docker Desktop enables users to build and share isolated, *containerised* applications on either a Mac or Windows environment. With 15 million monthly users, it's the most popular developer tool on the planet. Hence, **MirageOS networking libraries transparently handle the traffic of millions of containers**, simplifying many developers' experience.
+
+Docker initially started as an easy-to-use packaging of Linux kernels isolation primitives, such as [namespaces](https://en.wikipedia.org/wiki/Linux_namespaces) or [cgroups](https://en.wikipedia.org/wiki/Cgroups). Docker made it convenient for developers to encapsulate applications inside containers, protecting the rest of the system from potential bugs, moving them more easily between machines, and sharing them with other developers. Unfortunately, one of the challenges of running Docker on macOS or Windows is that these Linux primitives are unavailable. Docker Desktop goes to great lengths to emulate them without modifying the user experience of running containers. It runs a (light) Linux VM to host the Docker daemon with additional "glue" helpers to connect the Docker client, running on the host, to that VM. While the Linux VM can run Linux containers, it cannot directly access network resources like internal registries for VPNs specifically configured for forwarding traffic on the host. 
```

Comment:
maybe "Linux kernels isolation primitives" -> "the Linux kernel's isolation primitives"